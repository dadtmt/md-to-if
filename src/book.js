import * as R from 'ramda'
import { snakeCase } from 'change-case'

export const getSceneName = R.pipe(
  R.propOr([], 'content'),
  R.head,
  R.propOr('unnamed', 'content'),
  snakeCase
)

export const splitByScene = R.pipe(
  R.converge(R.prepend, [
    R.head,
    R.pipe(
      R.tail,
      R.splitWhen(
        R.where({
          type: R.equals('heading'),
          level: R.equals(2),
        })
      )
    ),
  ]),
  R.zipObj(['heading', 'content', 'sourceLeft']),
  ({ heading, content, sourceLeft }) => ({
    scene: {
      name: getSceneName(heading),
      content: [heading, ...content],
    },
    sourceLeft,
  })
)

const book = (source, scenes = []) => {
  const { scene, sourceLeft } = splitByScene(source)
  return R.isEmpty(sourceLeft)
    ? [...scenes, scene]
    : book(sourceLeft, [...scenes, scene])
}

export default book
