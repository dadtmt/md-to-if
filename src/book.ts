import * as R from 'ramda'
import { snakeCase } from 'change-case'

// Content -> String
export const getSceneName = R.pipe(
  R.propOr([], 'content'),
  R.head,
  R.propOr('unnamed', 'content'),
  snakeCase
)

// int -> [Content] -> [[Content],[Content]]
const splitByHeading = level =>
  R.splitWhen(
    R.where({
      type: R.equals('heading'),
      level: R.equals(level),
    })
  )

const splitActions = (level, contentList, actionList = []) => {
  const [headOfContent, ...tailOfContent] = contentList
  const [actionContent, restOfContent] = splitByHeading(level)(tailOfContent)
  const { content, actions } =
    level <= 6
      ? splitContentAndActions(level + 1)(actionContent)
      : {
          content: [],
          actions: [],
        }
  return !R.isEmpty(restOfContent)
    ? splitActions(level, restOfContent, [
        ...actionList,
        {
          name: getSceneName(headOfContent),
          sceneContent: [headOfContent, ...content],
          actions,
        },
      ])
    : actionList
}

// [Content] -> { content: [Content], actions: [Scene] }
export const splitContentAndActions = (level: number) => contentAndActions => {
  const [content, actions] = splitByHeading(level)(contentAndActions)
  return { content, actions: splitActions(level, actions) }
}

// [Content] -> { scene: Scene, content: [Content] }
export const splitByScene = (level: number): Function =>
  R.pipe(
    R.converge(R.prepend, [R.head, R.pipe(R.tail, splitByHeading(level))]),
    R.zipObj(['heading', 'content', 'sourceLeft']),
    ({ content, ...rest }) => ({
      content,
      ...rest,
    }),
    ({ heading, content, sourceLeft }) => ({
      scene: {
        name: getSceneName(heading),
        sceneContent: [heading, ...content],
      },
      sourceLeft,
    })
  )

// [Content], [Scene] -> [Scene]
const book = (source, scenes = []) => {
  const { scene, sourceLeft } = splitByScene(2)(source)
  return R.isEmpty(sourceLeft)
    ? [...scenes, scene]
    : book(sourceLeft, [...scenes, scene])
}

export default book
