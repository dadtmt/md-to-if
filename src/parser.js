import SimpleMarkdown from 'simple-markdown'

// String -> [String]
export const matchBraces = source => /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = source => /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = source => /^\|([\s\S]*?)\]/.exec(source)

const command = {
  order: 0,
  match: matchBraces,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

// String -> [Content]
const parser = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent,
})

export default parser
