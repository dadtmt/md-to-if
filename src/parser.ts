import SimpleMarkdown, { SingleASTNode } from 'simple-markdown'

// String -> [String]
export const matchBraces = (source: string): RegExpExecArray | null =>
  /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = (source: string): RegExpExecArray | null =>
  /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = (source: string): RegExpExecArray | null =>
  /^\|([\s\S]*?)\]/.exec(source)

const parse: (capture: any, parse: any, state: any) => any = (
  capture,
  parse,
  state
) => ({ content: parse(capture[1], state) })

const command = {
  order: 0,
  match: matchBraces,
  parse
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse
}

// String -> [Content]
const parser: (source: string) => SingleASTNode[] = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent
})

export default parser
