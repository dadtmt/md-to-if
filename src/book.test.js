import parser from './parser'
import book, { splitByScene, getSceneName } from './book'
import adventure from './adventure.md.js'

const lastSceneMd = `## Final landing

congrats you made it!!!
    `

describe('getSceneName', () => {
  it('get scene name', () => {
    const scene = parser(lastSceneMd)
    expect(getSceneName(scene)).toMatchSnapshot()
  })
})
describe('splitByScene', () => {
  it('split the first scene', () => {
    expect(splitByScene(parser(adventure))).toMatchSnapshot()
  })
  it('split the last scene', () => {
    expect(splitByScene(parser(lastSceneMd))).toMatchSnapshot()
  })
})

describe('book', () => {
  it('sort by scene', () => {
    expect(book(parser(adventure))).toMatchSnapshot()
  })
})
