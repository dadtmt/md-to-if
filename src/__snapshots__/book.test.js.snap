// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book sort by scene 1`] = `
Array [
  Object {
    "name": "choose_your_story",
    "sceneContent": Array [
      Object {
        "content": Array [
          Object {
            "content": "Choose your story",
            "type": "text",
          },
        ],
        "level": 1,
        "type": "heading",
      },
      Object {
        "content": Array [
          Object {
            "content": "This is a test adventure",
            "type": "text",
          },
          Object {
            "content": ", welcome",
            "type": "text",
          },
          Object {
            "content": ".",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
  },
  Object {
    "name": "the_space_ship",
    "sceneContent": Array [
      Object {
        "content": Array [
          Object {
            "content": "The space ship",
            "type": "text",
          },
        ],
        "level": 2,
        "type": "heading",
      },
      Object {
        "content": Array [
          Object {
            "content": "You are the brave pilot of this ship",
            "type": "text",
          },
          Object {
            "content": ".
The trip is long and boring",
            "type": "text",
          },
          Object {
            "content": ".",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": "Are you going to ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": "get food",
                "type": "text",
              },
            ],
            "target": "/cantina",
            "title": undefined,
            "type": "link",
          },
          Object {
            "content": " or to ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": "get a rest",
                "type": "text",
              },
            ],
            "target": "/bedroom",
            "title": undefined,
            "type": "link",
          },
        ],
        "type": "paragraph",
      },
    ],
  },
  Object {
    "name": "cantina",
    "sceneContent": Array [
      Object {
        "content": Array [
          Object {
            "content": "Cantina",
            "type": "text",
          },
        ],
        "level": 2,
        "type": "heading",
      },
      Object {
        "content": Array [
          Object {
            "content": " show playedCount ",
            "type": "text",
          },
        ],
        "type": "command",
      },
      Object {
        "content": Array [
          Object {
            "content": " time you come here",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": "The cantina is full with your crew having a good time ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " test playedCount equals val 1 ",
                "type": "text",
              },
            ],
            "type": "command",
          },
          Object {
            "content": " ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": "trueCase",
                "type": "text",
              },
            ],
            "type": "trueCaseContent",
          },
          Object {
            "content": Array [
              Object {
                "content": "falseCase",
                "type": "text",
              },
            ],
            "type": "falseCaseContent",
          },
          Object {
            "content": " ",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": "They are singing ",
            "type": "text",
          },
          Object {
            "content": "'oh my captain",
            "type": "text",
          },
          Object {
            "content": "' for the ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " show playedCount ",
                "type": "text",
              },
            ],
            "type": "command",
          },
          Object {
            "content": " times",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": "You pick up a sandwich an put it in your bag ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " set bag sandwich val 1 ",
                "type": "text",
              },
            ],
            "type": "command",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": "You find some pickles ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " set bag pickles val roll d10 ",
                "type": "text",
              },
            ],
            "type": "command",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": "You have ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " show bag sandwich ",
                "type": "text",
              },
            ],
            "type": "command",
          },
          Object {
            "content": " with you",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": "You have ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " show bag pickles ",
                "type": "text",
              },
            ],
            "type": "command",
          },
          Object {
            "content": " with you",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": " test playedCount equals val 1 ",
            "type": "text",
          },
        ],
        "type": "command",
      },
      Object {
        "content": Array [
          Object {
            "content": " ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " This is your first time here
    Will you come back ",
                "type": "text",
              },
              Object {
                "content": "? ",
                "type": "text",
              },
            ],
            "type": "trueCaseContent",
          },
          Object {
            "content": Array [
              Object {
                "content": " Should not you go somewhere else ",
                "type": "text",
              },
              Object {
                "content": "?
or what ",
                "type": "text",
              },
              Object {
                "content": "? ",
                "type": "text",
              },
            ],
            "type": "falseCaseContent",
          },
          Object {
            "content": " ",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
  },
  Object {
    "name": "bedroom",
    "sceneContent": Array [
      Object {
        "content": Array [
          Object {
            "content": "Bedroom",
            "type": "text",
          },
        ],
        "level": 2,
        "type": "heading",
      },
      Object {
        "content": Array [
          Object {
            "content": "Crazy major droid shoots you ",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": " describe droid ",
            "type": "text",
          },
          Object {
            "align": Array [
              null,
              null,
              null,
              null,
            ],
            "cells": Array [
              Array [
                Array [
                  Object {
                    "content": "roll D6",
                    "type": "text",
                  },
                ],
                Array [
                  Object {
                    "content": "33",
                    "type": "text",
                  },
                ],
                Array [
                  Object {
                    "content": "24",
                    "type": "text",
                  },
                ],
                Array [
                  Object {
                    "content": "41",
                    "type": "text",
                  },
                ],
              ],
            ],
            "header": Array [
              Array [
                Object {
                  "content": "CC",
                  "type": "text",
                },
              ],
              Array [
                Object {
                  "content": "CT",
                  "type": "text",
                },
              ],
              Array [
                Object {
                  "content": "F",
                  "type": "text",
                },
              ],
              Array [
                Object {
                  "content": "E",
                  "type": "text",
                },
              ],
            ],
            "type": "table",
          },
        ],
        "type": "command",
      },
      Object {
        "type": "newline",
      },
      Object {
        "content": Array [
          Object {
            "content": "You roll a dice it gives ",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " show roll d10 ",
                "type": "text",
              },
            ],
            "type": "command",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": "The droid try to shoot you
",
            "type": "text",
          },
          Object {
            "content": Array [
              Object {
                "content": " test d100 lte val droid CT ",
                "type": "text",
              },
            ],
            "type": "command",
          },
          Object {
            "content": Array [
              Object {
                "content": " you are shot ",
                "type": "text",
              },
            ],
            "type": "trueCaseContent",
          },
          Object {
            "content": Array [
              Object {
                "content": " You luckyly escape ",
                "type": "text",
              },
            ],
            "type": "falseCaseContent",
          },
        ],
        "type": "paragraph",
      },
    ],
  },
]
`;

exports[`splitByScene split the first scene 1`] = `
Object {
  "scene": Object {
    "name": "choose_your_story",
    "sceneContent": Array [
      Object {
        "content": Array [
          Object {
            "content": "Choose your story",
            "type": "text",
          },
        ],
        "level": 1,
        "type": "heading",
      },
      Object {
        "content": Array [
          Object {
            "content": "This is a test adventure",
            "type": "text",
          },
          Object {
            "content": ", welcome",
            "type": "text",
          },
          Object {
            "content": ".",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
  },
  "sourceLeft": Array [
    Object {
      "content": Array [
        Object {
          "content": "The space ship",
          "type": "text",
        },
      ],
      "level": 2,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": "You are the brave pilot of this ship",
          "type": "text",
        },
        Object {
          "content": ".
The trip is long and boring",
          "type": "text",
        },
        Object {
          "content": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "Are you going to ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": "get food",
              "type": "text",
            },
          ],
          "target": "/cantina",
          "title": undefined,
          "type": "link",
        },
        Object {
          "content": " or to ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": "get a rest",
              "type": "text",
            },
          ],
          "target": "/bedroom",
          "title": undefined,
          "type": "link",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "Cantina",
          "type": "text",
        },
      ],
      "level": 2,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": " show playedCount ",
          "type": "text",
        },
      ],
      "type": "command",
    },
    Object {
      "content": Array [
        Object {
          "content": " time you come here",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "The cantina is full with your crew having a good time ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " test playedCount equals val 1 ",
              "type": "text",
            },
          ],
          "type": "command",
        },
        Object {
          "content": " ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": "trueCase",
              "type": "text",
            },
          ],
          "type": "trueCaseContent",
        },
        Object {
          "content": Array [
            Object {
              "content": "falseCase",
              "type": "text",
            },
          ],
          "type": "falseCaseContent",
        },
        Object {
          "content": " ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "They are singing ",
          "type": "text",
        },
        Object {
          "content": "'oh my captain",
          "type": "text",
        },
        Object {
          "content": "' for the ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " show playedCount ",
              "type": "text",
            },
          ],
          "type": "command",
        },
        Object {
          "content": " times",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "You pick up a sandwich an put it in your bag ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " set bag sandwich val 1 ",
              "type": "text",
            },
          ],
          "type": "command",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "You find some pickles ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " set bag pickles val roll d10 ",
              "type": "text",
            },
          ],
          "type": "command",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "You have ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " show bag sandwich ",
              "type": "text",
            },
          ],
          "type": "command",
        },
        Object {
          "content": " with you",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "You have ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " show bag pickles ",
              "type": "text",
            },
          ],
          "type": "command",
        },
        Object {
          "content": " with you",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": " test playedCount equals val 1 ",
          "type": "text",
        },
      ],
      "type": "command",
    },
    Object {
      "content": Array [
        Object {
          "content": " ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " This is your first time here
    Will you come back ",
              "type": "text",
            },
            Object {
              "content": "? ",
              "type": "text",
            },
          ],
          "type": "trueCaseContent",
        },
        Object {
          "content": Array [
            Object {
              "content": " Should not you go somewhere else ",
              "type": "text",
            },
            Object {
              "content": "?
or what ",
              "type": "text",
            },
            Object {
              "content": "? ",
              "type": "text",
            },
          ],
          "type": "falseCaseContent",
        },
        Object {
          "content": " ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "Bedroom",
          "type": "text",
        },
      ],
      "level": 2,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": "Crazy major droid shoots you ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": " describe droid ",
          "type": "text",
        },
        Object {
          "align": Array [
            null,
            null,
            null,
            null,
          ],
          "cells": Array [
            Array [
              Array [
                Object {
                  "content": "roll D6",
                  "type": "text",
                },
              ],
              Array [
                Object {
                  "content": "33",
                  "type": "text",
                },
              ],
              Array [
                Object {
                  "content": "24",
                  "type": "text",
                },
              ],
              Array [
                Object {
                  "content": "41",
                  "type": "text",
                },
              ],
            ],
          ],
          "header": Array [
            Array [
              Object {
                "content": "CC",
                "type": "text",
              },
            ],
            Array [
              Object {
                "content": "CT",
                "type": "text",
              },
            ],
            Array [
              Object {
                "content": "F",
                "type": "text",
              },
            ],
            Array [
              Object {
                "content": "E",
                "type": "text",
              },
            ],
          ],
          "type": "table",
        },
      ],
      "type": "command",
    },
    Object {
      "type": "newline",
    },
    Object {
      "content": Array [
        Object {
          "content": "You roll a dice it gives ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " show roll d10 ",
              "type": "text",
            },
          ],
          "type": "command",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "The droid try to shoot you
",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " test d100 lte val droid CT ",
              "type": "text",
            },
          ],
          "type": "command",
        },
        Object {
          "content": Array [
            Object {
              "content": " you are shot ",
              "type": "text",
            },
          ],
          "type": "trueCaseContent",
        },
        Object {
          "content": Array [
            Object {
              "content": " You luckyly escape ",
              "type": "text",
            },
          ],
          "type": "falseCaseContent",
        },
      ],
      "type": "paragraph",
    },
  ],
}
`;

exports[`splitByScene split the last scene 1`] = `
Object {
  "scene": Object {
    "name": "final_landing",
    "sceneContent": Array [
      Object {
        "content": Array [
          Object {
            "content": "Final landing",
            "type": "text",
          },
        ],
        "level": 2,
        "type": "heading",
      },
      Object {
        "content": Array [
          Object {
            "content": "congrats you made it",
            "type": "text",
          },
          Object {
            "content": "!",
            "type": "text",
          },
          Object {
            "content": "!",
            "type": "text",
          },
          Object {
            "content": "!",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
  },
  "sourceLeft": Array [],
}
`;
